### ISSUE TITLE
Will this build for Hololens on 0.14393?

### ISSUE THREAD

**Issue by utekai** [xamarin/urho#324]
Having difficulty building and deploying successfully to Hololens on 0.14393. 

Are there specific build commands needed? Or what is the secret sauce to get it to work?

Using ...
1.  MakeWindows.bat x86 Debug 2017 DirectX/OpenGL (tried both, but think DirectX is correct)
2. MakeSharpReality.bat x86 Debug 2017

Then open in Visual Studio and modify ...
1.  UrhoSharp.SharpReality c++ to use 0.14393
2. UrhoSharp.SharpReality c# to use 0.14393 and modify json file to correct the framework
3. Playgrounds.SharpReality to use 0.14393 and modify json file
4. Remove and add back into Playgrounds.SharpReality project the mono-holourho.dll generated when compiling the native c++ UrhoSharp.SharpReality project (as the default path is for the x64 version which is not generated)

Any combination of steps results in ....

System.DllNotFoundException: 'Unable to load DLL 'mono-holourho': The specified module could not be found. (Exception from HRESULT: 0x8007007E)'

This in particular happens on the call in the UrhoAppView.cs file in UrhoSharp.SharpReality c# project ...

InitializeSpace();

Declared as
[DllImport(Consts.NativeImport, CallingConvention = CallingConvention.Cdecl)]
static extern void InitializeSpace();

And the Consts.NativeImport is a simple If on UWP_HOLO which specifies the mono-holourho dll.

**Comment by Noemata** [comment]
I had this same issue.  I should have written down the steps I went through to resolve this.  I vaguely remember adjusting the code that loads the DLL.  Are you building from the latest branch?  Perhaps we can help each other on this.  I'm not having any luck with the current branch.

**Comment by utekai** [comment]
Yes, using latest branch. But it's been setup for x64 and a Win10 UWP framework not available yet on Hololens. the 15083 branch, and hololens is still on 14393. 

Suspect or broken areas:

1. material.SetTechnique (memory access violation, but not sure if it's this call, or this just exposes another error, as memory access voilations start appearing on any SharpReality version higher than 1.5.22)
2. Sound playback via media player ... 
3. Spatial Mapping
4. General instability, likely related to #1.

As for loading the DLL, in particular mono-holourho.dll, this is problematic. It seems to be a dll needed by this dll that is causing the issue though. 

Currently trying to simplify the structure to figure it out.

Not sure what you mean by 'code that loads the dll'.

For instance, the call to 'InitializeSpace' is just declared as an extern and then called directly. From UrhoAppView.cs ... not sure what to adjust ...

[DllImport(Consts.NativeImport, CallingConvention = CallingConvention.Cdecl)]
static extern void InitializeSpace();

**Comment by utekai** [comment]
Alternatively, have also been just loading the previous SharpReality releases from nuget trying to narrow down particular changes that could be the cause, but not getting anywhere.

1.5.22 is stable, but not releasable. Anything above 1.5.22 starts throwing Memory Access Violations, in particular on the material.SetTechnique calls.

**Comment by utekai** [comment]
Tried branch release-1.5 and get same result, can't find mono-holourho.dll (but the dll is there, it's either not in correct format or another dll is required by this one) ... fewer changes needed though, as most projects are still setup with the correct windows 10 framework for hololens.

Playgrounds.SharpReality runs fine with the nuget SharpReality, but unable to link directly to the project UrhoSharp.SharpReality and get it to run for debugging purposes.

**Comment by utekai** [comment]
The crash happens in the file twinapi.appcore.dll, a System file.

I simplified by adding a new project that has all the files of Urho.SharpReality and thus calls directly to the dll.

Same as with the more complex path.  Perhaps time to load system symbols.

**Comment by utekai** [comment]
Thanks for these clues ... it's working now ... here's the algorithm followed:

1. using a GIT console window download the Urho project ... with this command ....

git clone https://github.com/xamarin/urho.git --recursive

2. using the Developer command prompt for VS2017, on a system with CMake installed ...

MakeSharpReality.bat x86 Release 2017 DirectX

3. Open the UrhoSolution ... with visual studio 2017 ... and modify one project

Close the project for Playgrounds.SharpReality and view code to open the csproj file.
Modify the link for mono-holourho.dll to use the Win32 folder instead of x64 folder

4. Modify the platform to 0.14399 on three projects in project settings

Playgrounds.SharpReality
Urho.SharpReality
UrhoSharp.SharpReality

5. Modify the project.json  to use the .14393 framework on these two projects
Urho.SharpReality
Playgrounds.SharpReality

6. Set Playgrounds.SharpReality as the Startup project
7. Modify the VS setting on toolbar from AnyCPU to x86 for release and add a remote machine (to hololens)

8. Run without debugging

**Comment by davejmurphy** [comment]
Glad that's working. 

Are you able to run the Playgrounds.SharpReality while VS toolbar is set to Debug, and then launch it with the debugger?

**Comment by utekai** [comment]
Regarding spatial mapping, the SpatialMappingTests file in Playgrounds.SharpReality overrides the call to adjust it as you mention, so it works in that file and shows a workaround for the existing bug.

**Comment by Noemata** [comment]
Clearly some folks have a different set of sources.  To get this to build I had to muck around more as shown below.  I have no clue how others were able to bypass this:

Modifications:
D:\Build\urho\Urho3D\Source\Source\Urho3D\IO\Log.cpp
Line 84:
#if !defined(__ANDROID__) && !defined(IOS) && !defined(TVOS)) && !defined(UWP)
to
#if !defined(__ANDROID__) && !defined(IOS) && !defined(TVOS) && !defined(UWP)

D:\Build\urho\Urho3D\Source\Source\Urho3D\IO\FileSystem.cpp
Line 710:
	programDir_ = AddTrailingSlash(SDL_UWP_GetResourceDir());
to
	String programDir_ = AddTrailingSlash(SDL_UWP_GetResourceDir());

D:\Build\urho\Urho3D\Source\Source\Urho3D\Core\ProcessUtils.cpp
Line 391:
    //int flags = fcntl(STDIN_FILENO, F_GETFL);
    //fcntl(STDIN_FILENO, F_SETFL, flags | O_NONBLOCK);
    //for (;;)
    //{
    //    int ch = fgetc(stdin);
    //    if (ch >= 0 && ch != '\n')
    //        ret += (char)ch;
    //    else
    //        break;
    //}
